### PR TITLE
pkg/conf: Add deadlock detector for frontend

### PR DESCRIPTION
**Pull request by ggilmore**
Many packages that we have use `conf.Watch` or `conf.Get` in their `init` functions. Such packages are included in `frontend`'s dependencies. Since `frontend` is responsible for serving the configuration to everyone else, it's possible to get into a situation where one of frontend's dependencies' `init` functions will block forever since `frontend` hasn't been initialized yet (and can't be until its dependencies have been initialized). 

The solution is to make sure that calls to `pkg/conf` never block the `init` function. 

This PR: 

1) Adds a detector that will `panic` if you're trying to call `conf.Get` or `conf.Watch` before `frontend` has been initialized 
2) Fixes a couple deadlocks due to these issues

### CODE REVIEW

File: pkg/conf/conf.go [hunk 1]
```diff
@@ -72,20 +83,41 @@ func init() {
...
-			configFilePath = "/etc/sourcegraph/config.json"
-		}
-		globals.ConfigurationServerFrontendOnly = confserver.NewServer(configFilePath)
+	// The default client is started in InitConfigurationServerFrontendOnly in
+	// the case of server mode.
+	if mode == modeClient {
+		close(configurationServerFrontendOnlyInitialized)
+
+		go defaultClient.continuouslyUpdate()
+
```
Comment (slimsag) [inline pkg/conf/conf.go]: Remove blank line here

File: pkg/conf/store.go [hunk 2]
```diff
@@ -113,6 +113,16 @@ func (s *Store) MaybeUpdate(rawConfig string) (UpdateResult, error) {
 // WaitUntilInitialized blocks and only returns to the caller once the store
 // has initialized with a syntactically valid configuration file (via MaybeUpdate() or Mock()).
 func (s *Store) WaitUntilInitialized() {
+	mode := getMode()
+	if mode == modeServer {
+		select {
+		case <-configurationServerFrontendOnlyInitialized:
+			// Configuration server is initialized or this is not running in server mode.
```
Comment (slimsag) [inline pkg/conf/store.go]: How is this possible? This conditional says `mode == modeServer`, so this cannot occur in non-server mode. The comment seems misleading.

File: cmd/frontend/db/phabricator.go [hunk 3]
```diff
@@ -18,24 +19,37 @@ type errPhabricatorRepoNotFound struct {
...
-		phabricatorRepos.Set(func() interface{} {
-			repos := map[api.RepoName]*types.PhabricatorRepo{}
-			for _, config := range conf.Get().Phabricator {
-				for _, repo := range config.Repos {
-					repos[api.RepoName(repo.Path)] = &types.PhabricatorRepo{
-						Name:     api.RepoName(repo.Path),
-						Callsign: repo.Callsign,
-						URL:      config.Url,
+	go func() {
+		conf.WaitUntilConfigurationServerInitialized()
```
Comment (slimsag) [inline cmd/frontend/db/phabricator.go]: I don't understand why we need `WaitUntilConfigurationServerInitialized`. This seems like a detail that `conf.Get()` should be handling and should not be exposed outside of `pkg/conf`.

**Comment by beyang** [comment]
@ggilmore @slimsag How does a client of the `conf` package know if they must call `conf.Watch` in a go routine inside `init`? I don't have a deep understanding of the code here, but it seems like a client of `conf` now needs to understand some details about the concurrency of the implementation?

Here's my understanding of the deadlock: Some `init` functions call `conf.Get` and `conf.Get` depends on the frontend service fully spinning up (and specifically connecting to the DB where the conf source of truth now is). It seems to me that there must be a better way around this that doesn't involve clients of `conf` having to think about whether they should call `conf` functions inside of a goroutine or not (and add additional synchronization themselves).

Proposed solution (based on my limited knowledge of this code): why not get rid of `conf.Get` and only expose `conf.Watch`? That way, initializers that depend on `conf` will get invoked whenever when the `conf` package determines it is ready, not when the actual `init` functions are called. Another idea is a `conf.Ready` function that exposes the same interface as `conf.Watch` but only gets called once on initial `conf` readiness (if clients don't want to subscribe to changes in conf).

Basically, I'd like to avoid any synchronization code that depends on behavior defined both inside and outside `conf`. Concurrency should be well encapsulated. Otherwise, it seems like a huge headache waiting to happen.

**Comment by slimsag** [comment]
- `conf.Watch` blocks. If you use it, it is guaranteed that the function you pass in is executed at least once before it returns. Multiple callers rely on this behavior today.
- Every caller of `conf.Watch` should be doing so inside of a goroutine. `conf.Watch` blocks, and you should not be doing heavy work inside of `init` functions.
- If you do not do so in the frontend, you will receive a panic. If you do not do so in another service, your code should not have passed code review but it will work and you won't receive a panic.
- This has nothing to do with `conf.Get` vs. `conf.Watch`, because any time you use `conf.Watch`, you are inherently using `conf.Get` (because `conf.Watch` is synchronous).

> I don't have a deep understanding of the code here, but it seems like a client of conf now needs to understand some details about the concurrency of the implementation?

No, they do not need any such understanding.

If you use `conf.Watch`, you need to do so in a asynchronous manner. If you fail to do so, you may receive a panic.

This is not just limited to `conf.Watch` either: in general, all work done inside of `init` should be asynchronous because you are blocking other `init` functions from running. It just so happens that `conf.Watch` somewhat enforces this (because otherwise there would be a deadlock).

**Comment by beyang** [comment]
@slimsag Could you update the docstring of `conf.Watch` to reflect this? Currently, it reads,

```
// Watch calls the given function in a separate goroutine whenever the
// configuration has changed. The new configuration can be received by calling
// conf.Get.
//
// Before Watch returns, it will invoke f to use the current configuration.
//
// Watch is a wrapper around client.Watch.
 ```

* The fact that it "calls the given function in a separate goroutine" suggests non-blocking behavior.
* If it should *never* be called synchronously inside `init`, that should be mentioned.

**Comment by slimsag** [comment]
Yes
